trieste/acquisition/function/function.py

```diff
@@ -468,6 +468,84 @@ def acquisition(x: TensorType) -> TensorType:
     return acquisition
 
 
+class FastConstraintsFeasibility(SingleModelAcquisitionBuilder[ProbabilisticModel]):
+    """
+    Uses the :func:`fast_constraints_feasibility` function to build a
+    probability of feasiblity acquisition function from the residuals of direct constraints defined
+    in the search space.
+    """
+
+    def __init__(
+        self,
+        search_space: SearchSpace,
+        smoothing_function: Optional[Callable[[TensorType], TensorType]] = None,
+    ):
+        """
+        :param search_space: The global search space over which the feasibility of the constraints
+            is defined.
+        :param smoothing_function: The smoothing function used for constraints residuals. The
+            default is CDF of the Normal distribution with a small scale.
+        """
+        self._search_space = search_space
+        self._smoothing_function = smoothing_function
+
+    def __repr__(self) -> str:
+        """"""
+        return f"FastConstraintsFeasibility({self._search_space!r}, {self._smoothing_function!r})"
+
+    def prepare_acquisition_function(
+        self,
+        model: ProbabilisticModel,
+        dataset: Optional[Dataset] = None,
+    ) -> AcquisitionFunction:
+        """
+        :param model: Unused.
+        :param dataset: Unused.
+        :return: The function for probability of feasibility of constraints.
+        """
+        return fast_constraints_feasibility(self._search_space, self._smoothing_function)
+
+    def update_acquisition_function(
+        self,
+        function: AcquisitionFunction,
+        model: ProbabilisticModel,
+        dataset: Optional[Dataset] = None,
+    ) -> AcquisitionFunction:
+        """
+        :param function: The acquisition function to update.
+        :param model: Unused.
+        :param dataset: Unused.
+        :return: The function for probability of feasibility of constraints.
+        """
+        return function  # No need to update anything.
+
+
+def fast_constraints_feasibility(
+    search_space: SearchSpace,
+    smoothing_function: Optional[Callable[[TensorType], TensorType]] = None,
+) -> AcquisitionFunction:
+    """
+    :param search_space: The global search space over which the feasibility of the constraints
+        is defined.
+    :param smoothing_function: The smoothing function used for constraints residuals. The
+        default is CDF of the Normal distribution with a small scale.
+    :return: The function for probability of feasibility of constraints.
+    """
+
+    @tf.function
```

Comment:
Looking at the notebook, it looks like this is being retraced every loop, which is not ideal! I think the issue is because ExpectedConstrainedImprovement calls the constraint function with the query points, which change shape each time. We may be able to specify an input_signature here to handle this. If you can't quickly figure it out don't worry though: I'll raise an issue since I expect it's happening for the normal ProbabilityOfFeasibility too.

Reply:
Yes, `ProbabilityOfFeasibility` has that problem too. I did spend a little bit of time looking at it, but didn't try changing the signature.

Reply:
I've raised an issue, thanks: https://github.com/secondmind-labs/trieste/issues/661